tensorboard/plugins/hparams/hparams_plugin.py

```diff
@@ -90,6 +102,107 @@ def is_active(self):
     def frontend_metadata(self):
         return base_plugin.FrontendMetadata(element_name="tf-hparams-dashboard")
 
+    # ---- /download_data- -------------------------------------------------------
+    @wrappers.Request.application
+    def download_data_route(self, request):
+        response_format = request.args.get("format")
+        request_proto = _parse_request_argument(
+            request, api_pb2.ListSessionGroupsRequest
+        )
+        session_groups = list_session_groups.Handler(
+            self._context, request_proto
+        ).run()
+        experiment = get_experiment.Handler(self._context).run()
+        try:
+            body, mime_type = self.download_data_impl(
+                experiment, session_groups, response_format
+            )
+        except ValueError as e:
+            return http_util.Respond(
+                request=request,
+                content=str(e),
+                content_type="text/plain",
+                code=400,
+            )
+        return http_util.Respond(request, body, mime_type)
+
+    def download_data_impl(self, experiment, session_groups, response_format):
+        """Provides a response for downloading data for an HParams table.
+
+        Args:
+          experiment: Experiment proto.
+          session_groups: ListSessionGroupsResponse proto.
+          response_format: A string in the OutputFormat enum.
+
+        Raises:
+          ValueError: If the scalars plugin is not registered.
+
+        Returns:
+          2 entities:
+            - A JSON object response body.
+            - A mime type (string) for the response.
+        """
+        header = []
+        for hparam_info in experiment.hparam_infos:
+            header.append(hparam_info.display_name or hparam_info.name)
+
+        for metric_info in experiment.metric_infos:
+            header.append(metric_info.display_name or metric_info.name.tag)
+
+        rows = []
+
+        def _get_value(value):
+            if value.HasField("number_value"):
+                return value.number_value
+            if value.HasField("string_value"):
+                return value.string_value
+            if value.HasField("bool_value"):
+                return value.bool_value
+            return None
```

Comment:
Is this case meant to be reachable? If so, how—maybe comment? it’s not
obvious to me—and is `None` the right result? E.g., `latex_format(None)`
will raise a `TypeError`. If not, should we raise an `AssertionError` to
fail fast in case of a logic error?